### PR TITLE
infra(e2e): rework local E2E script

### PR DESCRIPTION
**Pull request by kelset**
## Summary

This is a long time coming effort to improve the situation around the local e2e script that in the release crew: the current bash-based script is quirky at best, and what you end up generating as a sample project is not really a true sample project. This is where this PR comes in: it migrates the flow from `./scripts/test-manual-e2e.sh` to `yarn test-e2e-local <options>`.

Here's the current shape of the options:

```sh
Options:
  --help          Show help                                            [boolean]
  --version       Show version number                                  [boolean]
  -t, --target      [choices: "RNTester", "RNTestProject"] [default: "RNTester"]
  -p, --platform                    [choices: "iOS", "Android"] [default: "iOS"]
  -h, --hermes                                         [boolean] [default: true]
```

The idea is to change it so that you can just run the script, and it will do that one specific thing "well", without the tester needing to do anything aside from actually testing the app once it's open.

Some of the key changes:
* tries to stick to the patterns of the other established *.js based scripts, in terms of tooling and approach (and even refactor parts that can be shared with other scripts) - like the android artifacts generation
* no need to start the android emulator on the side
* no need to start Metro on the side
* RNTester iOS will open up on the simulator (no Xcode open that then you need to press)

Things that still need work:
* see the #fixme and #todo in comments
* because we rely on exec, the output sent back is not formatted/shaped correctly so it's a bit more noisy/chaotic - but can't handle it right now because the package we use doesn't allow it - see https://github.com/shelljs/shelljs/issues/86

## Changelog

[Internal] [Changed] - Migrate bash E2E local testing script to new JS based command

## Test Plan

To test the script, you can run it passing the options showed above; this is the current situation:
* RNTester iOS Hermes ✅
* RNTester Android Hermes ✅
* RNTester iOS JSC ✅
* RNTester Android JSC ✅
* RNTestProject Android Hermes ✅
* RNTestProject iOS Hermes ✅
* RNTestProject Android JSC ✅
* RNTestProject iOS JSC ✅

### CODE REVIEW

**Comment by kelset** [comment]
update on solving the Hermes on iOS for RNTestProject: I managed to figure out the logic to make it build from source on the new project for when testing in main branch and from release branch.

BUT this caused this to happen:
Screenshot_2022-09-14_at_18 24 41

Basically, on the Android side everything gets build with the correct configuration, while on iOS not. This took a lot of digging, and a bit of fortune and "lucky timing", but basically now we know what's going on (it's the same issue @Kudo is trying to [address here](https://github.com/reactwg/react-native-new-architecture/discussions/68#discussioncomment-3652071)). Long story short, how hermes gets build on iOS has 3 different variations:
* there's `build-apple-framework.sh` in the Hermes repo, that runs [this Cmake command](https://github.com/facebook/hermes/blob/main/utils/build-apple-framework.sh#L61)
* there's `build-apple-framework.sh` in the React Native repo, that runs [this variation of the Cmake command](https://github.com/facebook/react-native/blob/main/sdks/hermes-engine/utils/build-apple-framework.sh#L51) that in particular has this extra option `-DHERMES_RELEASE_VERSION="for RN $(get_release_version)" \` that makes the right version of Hermes be generated (and displayed in a fresh new project)
* there's a variation of the above that is run in the RN CI, but it's all "manual" while the ones above are run through cocoapods. See here [the scripts](https://github.com/facebook/react-native/blob/main/.circleci/config.yml#L1044).

The core problem is that, within the `hermes-engine.podspec` contained in the RN repo (at `react-native/main/sdks/hermes-engine/`), there's a bit of logic that triggers that script [here](https://github.com/facebook/react-native/blob/main/sdks/hermes-engine/hermes-engine.podspec#L69) (through `./utils/build-ios-framework.sh`). The issue is that we all thought that that  `./utils/build-ios-framework.sh` would invoke the React native version of the scripts (since the podspec file lives right next to the `utils` folder) but, in reality, it doesn't. It just so happens that the Hermes repo has a root level `utils` folder which is (you guessed it) where the Hermes variation of those build scripts live.
So, when running the pod install command in a react-native project (build from source), it will go and download the hermes source code (since the `source[:git]` gets set) but then it will use the **hermes** variation of the `build-*.sh` scripts.

So we need to figure out how to get the pod process to build using the React Native variation of those scripts. @Kudo has proposed [a patch here](https://github.com/reactwg/react-native-new-architecture/discussions/68#discussioncomment-3654191) that would have some more explicit pathing - which seems a good workaround, but it doesn't seem to do the trick:
Screenshot 2022-09-15 at 16 31 27

So currently I'm freezing working on this until this build hermes issue is addressed.

**Comment by kelset** [comment]
update, with the latest updated patch proposed by Kudo, this works correctly now 🎉

Screenshot 2022-09-16 at 15 24 19

File: scripts/testing-utils.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  } catch (_error) {
+    return 'not_running';
+  }
+}
+
+// this is a very limited implementation of how this should work
+// literally, this is macos only
+// a more robust implementation can be found here:
+// https://github.com/react-native-community/cli/blob/7c003f2b1d9d80ec5c167614ba533a004272c685/packages/cli-platform-android/src/commands/runAndroid/index.ts#L195
+function launchPackagesInSeparateWindow() {
```
Comment (cortinico) [inline scripts/testing-utils.js:103]:
```suggestion
function launchPackagerInSeparateWindow() {
```

File: scripts/test-e2e-local-clean.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+// RNTestProject
+console.info('\n** Removing the RNTestProject folder **\n');
+exec('rm -rf /tmp/RNTestProject');
+
+// final clean up
+console.info('\n** Final git level wipe **\n');
+// clean unstaged changes from git
+exec('git checkout -- .');
+// remove all the untracked files
+exec('git clean -fdx');
```
Comment (cortinico) [inline scripts/test-e2e-local-clean.js:67]: The Android clean can be omitted if we run this.

File: scripts/test-e2e-local.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    // we do the android path here
+
+    launchAndroidEmulator();
+
+    if (argv.hermes) {
+      console.info(
+        "We're going to test the Hermes version of RNTester Android",
+      );
+      exec(
+        './gradlew :packages:rn-tester:android:app:installHermesDebug --quiet',
```
Comment (cortinico) [inline scripts/test-e2e-local.js:105]: I would string compose here to collapse lines 100-110

File: scripts/test-e2e-local.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    // if everything succeeded so far, we can launch Metro and the app
+    // start the Metro server in a separate window
+    launchPackagesInSeparateWindow();
+    // just to make sure that the Android up won't have troubles finding the Metro server
+    exec('adb reverse tcp:8081 tcp:8081');
+    // launch the app
+    exec(
+      'adb shell am start -n com.facebook.react.uiapp/com.facebook.react.uiapp.RNTesterActivity',
```
Comment (cortinico) [inline scripts/test-e2e-local.js:119]: Idea for future self: it should be possible to `yarn android` inside `packages/rn-tester`

File: scripts/test-e2e-local.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    launchPackagesInSeparateWindow();
+
+    // launch the app on iOS simulator
+    pushd('packages/rn-tester');
+    exec('npx react-native run-ios --scheme RNTester');
+    popd();
+  } else {
+    // we do the android path here
+
+    launchAndroidEmulator();
```
Comment (cortinico) [inline scripts/test-e2e-local.js:98]: Is it intentional that the Android emulator starts only for RN Tester?

**Review by cortinico — COMMENTED** [review]
Amazing stuff @kelset! Left minor comments around

File: scripts/test-e2e-local.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+const {generateAndroidArtifacts} = require('./release-utils');
+
+// setting this just 'cause this https://github.com/facebook/react-native/commit/0a3ca80af401654896cbc73e235711eef0f9b3c5
+// breaks a bunch of stuff. This should be enough.
+// TODO: follow up with Nicola in understanding the impact of the commit, and if we can remove this
+// also, remove the folder
+process.env.TMP_PUBLISH_DIR = fs.mkdtempSync(
+  path.join(os.tmpdir(), 'rn-publish-'),
+);
```
Comment (cortinico) [inline scripts/test-e2e-local.js:39]: cc @vincenzovitale

Comment (vincenzovitale) [reply scripts/test-e2e-local.js:39]: @kelset 
#34746 will prevent the setting of 
process.env.TMP_PUBLISH_DIR

File: scripts/test-e2e-local-clean.js (hunk 2 above)
Comment (kelset) [reply scripts/test-e2e-local-clean.js:67]: I see - but I kind of want to have it "spelled out" as a reference (= if someone checks out this script and sees that I use gradle CleanAll, they will know/think/consider it the better command for doing android-side cleaning)

File: scripts/test-e2e-local.js (hunk 4 above)
Comment (kelset) [reply scripts/test-e2e-local.js:113]: it might work actually, let me give it a try...

Comment (kelset) [reply scripts/test-e2e-local.js]: ok so for the future self; I tried replacing this with

```
    // launch the app on Android simulator
    pushd('packages/rn-tester');
    exec('npx react-native run-android');
    popd();
```

and it fails with `error Failed to install the app. Make sure you have the Android development environment set up: https://reactnative.dev/docs/environment-setup.`

so there's some custom stuff in rn-tester folder that we'd need to address for this to work. I think we can leave it for a future task

Comment (cortinico) [reply scripts/test-e2e-local.js:120]: > so there's some custom stuff in rn-tester folder that we'd need to address for this to work. I think we can leave it for a future task

Yeah, something to do in the future :+1:

**Comment by kelset** [comment]
@dmytrorykun just an headsup, I'm testing on a release branch and I've hit an issue... I'll let you know once I've fixed it

**Comment by kelset** [comment]
@dmytrorykun / @cortinico ok nvm the issue I was hitting was caused by me not having cherry-picked all the necessary changes back on top of 0.70 to test my script correctly; after a bit of investigation it's clear that it's tooo complicated to backport this script so we'll have to face potential/eventual issues during the 0.71 RC phase - but on paper everything should work fine 🤞😇

**Review by cortinico — APPROVED** [review]

**Comment by kelset** [comment]
(note to self: this page of the docs https://reactnative.dev/contributing/how-to-contribute-code#development-workflow needs to be updated once this is merged)

**Comment by cortinico** [comment]
Can we solve the merge conflict?

**Comment by kelset** [comment]
@cortinico done 👍

**Comment by cortinico** [comment]
Are we fine merging this?

**Pull request closed without merging**